### PR TITLE
Use case insensitive search in PostgreSQL

### PR DESCRIPTION
**Pull request by joelmellon**
Use the ILIKE operator to search columns ignoring case

### BEFORE - What was wrong? What was happening before this PR?
Searching in list views was case sensitive when using PostgreSQL leading to difficulties finding desired results.

### AFTER - What is happening after this PR?
Searches in list views return matching results regardless of their case in the database.

### How did you achieve that, in technical terms?
By using PostgreSQL's ILIKE operator as described here: https://www.postgresql.org/docs/8.3/functions-matching.html

### Is it a breaking change?
No, unless there are any fringe users running PostgreSQL and expecting their searches to be very strict.

##### Before
Perform a search that exactly matches the case in a column, see results.
Perform the same search but with text that varies from data in column, see no results.
##### After
Perform a search that exactly matches the case in a column, see results.
Perform the same search but with text that varies from data in column, see results.

#### Using other drivers, eg. MySQL, SQLite
See no change in behavior.

### CODE REVIEW

File: src/app/Library/CrudPanel/Traits/Search.php [hunk 1]
```diff
@@ -61,11 +62,18 @@ public function applySearchLogicForColumn($query, $column, $searchTerm)
 
         // sensible fallback search logic, if none was explicitly given
         if ($column['tableColumn']) {
+
```
Comment (joelmellon) [inline src/app/Library/CrudPanel/Traits/Search.php:65]:
```suggestion
```

**Comment by pxpm** [comment]
Hello @joelmellon thanks for the PR. 

I don't use Postgre myself, just wondering if there are other places where Backpack could see some improvement in places like this that you already noticed ? I say that because if we get to the point of checking the grammar in multiple places it would be better to start from scratch and just create some helper functions or whatever to get the "mutations" depending on the query grammar. 

Also if there are more operators that could mutate etc etc it's better to have a service class or something that get us the `$operators` in the current query grammar. I imagine like if you write `QueryOperation::like()`, it would automatically search the `like`, found it in `Mysql`, if the grammar is `pg` use the alias `ilike`, same thing otherway around. 

I mean, this is alot of thinkering and work to make it "perfect 99,9%", but it's only worth if there are a need for it, if it's just one or two cases we can do it like you did in this PR. 

Once again, thanks 🙏

**Comment by joelmellon** [comment]
@pxpm 

I’m new to PostgreSQL myself, so I only “fixed” this since I bumped into it during a transition from MySQL. I’ve yet to see other places that broke or had unexpected behavior.

I think 95%+ of the time Eloquent abstracts subtleties in grammar like this, so because Backpack is written so well, I think this type of thing will be rare.

It’s pretty surprising that Eloquent doesn’t offer a whereLike() method, but I’m going to postulate that differences in the way various engines handle case sensitivity is actually one of the reasons to omit such a common thing. (There are tons of whereLike() macros out there due to this, demonstrating the common need for it. Come to think of it, this could be a decent way to abstract this code if it comes up anywhere else.)

Anyway, I’m sure there are some other places that might need PostgreSQL specific syntax but I doubt there are enough to warrant any abstraction really. This PR didn’t feel particularly “clean” but my mind took the same path yours did, and this felt best.

Beyond this LIKE subtlety and my own super complicated queries, I don’t know enough about PostgreSQL to say what else could be weird. Nor do I know enough about Eloquent’s limitations at the edges of normal use.

For what it’s worth, PostgreSQL doesn’t seem to be all that popular, at least not as popular as I thought. It feels like a niche kind of thing in the PHP community at least. I started using it recently for a geography related project since it’s sort of the de facto GIS RDBMS engine, and not because I had to, but because I wanted to learn it. There is also a slightly bigger PHP community around it for this niche specifically, ie. more active GIS projects, than for MySQL, thorough MySQL offers a ton of the same functionality.

I always heard MySQL was behind the times, but we hear that about PHP still too. Meanwhile they both serve more websites than any other technologies.

TL;DR: I wouldn’t worry too much about PostgreSQL support. Laravel handles the majority of it.

**Comment by joelmellon** [comment]
I’ll just add that if you happen to find other places that could use this treatment and have a decent design pattern you’d like to see implemented, I’d be happy to contribute what I can.

**Comment by joelmellon** [comment]
Also, this is an issue for others: #1980

**Comment by pxpm** [comment]
Hello @joelmellon , me and @tabacitu already had a talk about this.

We decided that we are going to do it but in a different way to avoid conditionals.

What we have in the roadmap is adding `searchOperator` (`like` by default) as a config option in the `ListOperation`. That way developer can configure it like: `CRUD::setOperationSetting('searchOperator', 'ilike')` 

I will back to you with news very soon on this. 

Cheers

**Comment by pxpm** [comment]
Hey @joelmellon 

Thanks again for all your effort here. I've just submitted #4314 with the solution me and @tabacitu discussed. We think this leaves the possibility to also use `like` in postgre if you want to and solves your use-case too if all you want is use `ilike` in all ListOperations.

Sorry this didn't get merged, it was 100% helpfull in determining the proper solution.

I am going to close this since we have the PR open, if you happen to give it a try, please let us know in the PR thread 👍 

Cheers

**Comment by joelmellon** [comment]
@pxpm I really like your solution in #4314; it's so simple.

I'm also stoked you updated the config, thanks for taking the time to address this. 🙏

You guys rock! 🤘

**Comment by tabacitu** [comment]
I've just merged #4314 . It'll be tagged and available with a `composer update` on Monday the latest. Thank you for your patience @joelmellon 🙏
